File: src/planner/generate_grid.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <random>
#include <cmath>
#include <string>

// ========================== 参数配置 ==========================
const int GRID_SIZE = 256;           // 网格大小 256x256
const double OBSTACLE_RATIO = 0.2;  // 障碍物占比 (0.0 ~ 1.0)
const int MARGIN = 10;               // 起点/终点距边界的最小距离
const double MIN_DIST_RATIO = 0.3;   // 起点到终点的最小距离比例

// ========================== 随机数生成 ==========================
// 取消注释下面一行可固定随机种子，便于复现
// #define FIXED_SEED 42

// ========================== PPM 颜色定义 ==========================
struct Color {
    int r, g, b;
};

const Color WHITE    = {255, 255, 255};  // 空闲
const Color BLACK    = {0,   0,   0  };  // 障碍
const Color BLUE     = {0,   0,   255};  // 起点
const Color RED      = {255, 0,   0  };  // 终点

// ========================== 网格生成 ==========================

// 在 grid 上填充一个矩形障碍物，跳过起点/终点区域（如果已设置）
void fillRect(std::vector<std::vector<int>>& grid,
              int r1, int c1, int r2, int c2) {
    for (int r = r1; r <= r2; r++) {
        for (int c = c1; c <= c2; c++) {
            grid[r][c] = 1;
        }
    }
}

void generateGrid(std::vector<std::vector<int>>& grid, std::mt19937& rng) {
    grid.assign(GRID_SIZE, std::vector<int>(GRID_SIZE, 0));

    int total_cells = GRID_SIZE * GRID_SIZE;
    int target_obs = static_cast<int>(total_cells * OBSTACLE_RATIO);
    int placed = 0;

    // 矩形尺寸范围
    std::uniform_int_distribution<int> w_dist(4, 30);  // 宽度
    std::uniform_int_distribution<int> h_dist(4, 30);  // 高度
    std::uniform_int_distribution<int> pos_dist(0, GRID_SIZE - 1);

    // 随机放置矩形块
    while (placed < target_obs) {
        int w = w_dist(rng);
        int h = h_dist(rng);
        int r = pos_dist(rng);
        int c = pos_dist(rng);

        // 裁剪到边界内
        int r_end = std::min(r + h - 1, GRID_SIZE - 1);
        int c_end = std::min(c + w - 1, GRID_SIZE - 1);

        // 检查该区域是否全为空（避免重叠太多）
        int area = (r_end - r + 1) * (c_end - c + 1);
        int already_filled = 0;
        for (int rr = r; rr <= r_end; rr++)
            for (int cc = c; cc <= c_end; cc++)
                if (grid[rr][cc] == 1) already_filled++;

        // 如果重叠超过 30%，跳过
        if (already_filled > area * 0.3) continue;

        fillRect(grid, r, c, r_end, c_end);
        placed += (area - already_filled);
    }

    // 额外添加一些细长的"墙壁"（横向 + 纵向），让地图更有结构感
    std::uniform_int_distribution<int> wall_len(20, 80);
    std::uniform_int_distribution<int> thin_pos(5, GRID_SIZE - 5);
    int walls = GRID_SIZE / 16;  // ~16 面墙

    for (int i = 0; i < walls; i++) {
        int len = wall_len(rng);
        int r = thin_pos(rng);
        int c = thin_pos(rng);

        if (i % 2 == 0) {
            // 横向墙 (厚度 1~2)
            int thick = (rng() % 2) + 1;
            int c_end = std::min(c + len, GRID_SIZE - 1);
            int r_end = std::min(r + thick - 1, GRID_SIZE - 1);
            fillRect(grid, r, c, r_end, c_end);
        } else {
            // 纵向墙 (厚度 1~2)
            int thick = (rng() % 2) + 1;
            int r_end = std::min(r + len, GRID_SIZE - 1);
            int c_end = std::min(c + thick - 1, GRID_SIZE - 1);
            fillRect(grid, r, c, r_end, c_end);
        }
    }
}

// ========================== 起点/终点生成 ==========================
struct Point {
    int row, col;
};

void generateStartGoal(const std::vector<std::vector<int>>& grid,
                       Point& start, Point& goal,
                       std::mt19937& rng) {
    // 收集符合条件的空闲格子
    std::vector<Point> free_cells;
    for (int r = MARGIN; r < GRID_SIZE - MARGIN; r++) {
        for (int c = MARGIN; c < GRID_SIZE - MARGIN; c++) {
            if (grid[r][c] == 0) {
                free_cells.push_back({r, c});
            }
        }
    }

    if (free_cells.size() < 2) {
        std::cerr << "错误: 空闲格子不足，无法放置起点和终点!" << std::endl;
        exit(1);
    }

    std::uniform_int_distribution<size_t> dist(0, free_cells.size() - 1);
    double min_dist = GRID_SIZE * MIN_DIST_RATIO;

    // 保证起点和终点有足够的间距
    do {
        size_t i1 = dist(rng);
        size_t i2 = dist(rng);
        while (i2 == i1) i2 = dist(rng);

        start = free_cells[i1];
        goal  = free_cells[i2];
    } while (std::hypot(start.row - goal.row, start.col - goal.col) < min_dist);
}

// ========================== 保存文本网格 ==========================
void saveGridText(const std::vector<std::vector<int>>& grid,
                  const Point& start, const Point& goal,
                  const std::string& filepath) {
    std::ofstream out(filepath);
    if (!out.is_open()) {
        std::cerr << "错误: 无法创建文件 " << filepath << std::endl;
        return;
    }

    out << "# A* Grid Map\n";
    out << "# size: " << GRID_SIZE << "x" << GRID_SIZE << "\n";
    out << "# start: (" << start.row << ", " << start.col << ")\n";
    out << "# goal:  (" << goal.row << ", " << goal.col << ")\n";
    out << "# 0=free, 1=obstacle\n";
    out << GRID_SIZE << "\n";
    out << start.row << " " << start.col << "\n";
    out << goal.row << " " << goal.col << "\n";

    for (int r = 0; r < GRID_SIZE; r++) {
        for (int c = 0; c < GRID_SIZE; c++) {
            out << grid[r][c];
            if (c < GRID_SIZE - 1) out << " ";
        }
        out << "\n";
    }

    out.close();
    std::cout << "网格数据已保存至: " << filepath << std::endl;
}

// ========================== 保存 PPM 图片 ==========================
void saveGridPPM(const std::vector<std::vector<int>>& grid,
                 const Point& start, const Point& goal,
                 const std::string& filepath) {
    std::ofstream out(filepath);
    if (!out.is_open()) {
        std::cerr << "错误: 无法创建文件 " << filepath << std::endl;
        return;
    }

    // PPM P3 (ASCII) header
    out << "P3\n";
    out << GRID_SIZE << " " << GRID_SIZE << "\n";
    out << "255\n";

    // 放大每个格子让图片更清晰 (pixel scale factor)
    const int SCALE = 3;  // 每个格子 3x3 像素，使 256x256 的图片不至于太小
    int img_w = GRID_SIZE * SCALE;
    int img_h = GRID_SIZE * SCALE;

    out.seekp(0);
    out << "P3\n";
    out << img_w << " " << img_h << "\n";
    out << "255\n";

    for (int r = 0; r < GRID_SIZE; r++) {
        for (int sr = 0; sr < SCALE; sr++) {
            for (int c = 0; c < GRID_SIZE; c++) {
                Color color;
                if (r == start.row && c == start.col) {
                    color = BLUE;
                } else if (r == goal.row && c == goal.col) {
                    color = RED;
                } else if (grid[r][c] == 1) {
                    color = BLACK;
                } else {
                    color = WHITE;
                }

                // 每个格子填充 SCALE 个像素
                for (int sc = 0; sc < SCALE; sc++) {
                    out << color.r << " " << color.g << " " << color.b << " ";
                }
            }
            out << "\n";
        }
    }

    out.close();
    std::cout << "图片已保存至: " << filepath << std::endl;
}

// ========================== 主函数 ==========================
int main() {
#ifdef FIXED_SEED
    std::mt19937 rng(FIXED_SEED);
    std::cout << "使用固定随机种子: " << FIXED_SEED << std::endl;
#else
    std::random_device rd;
    std::mt19937 rng(rd());
    std::cout << "使用随机设备生成种子" << std::endl;
#endif

    std::cout << "=== A* 网格地图生成 (C++) ===" << std::endl;
    std::cout << "网格大小: " << GRID_SIZE << "x" << GRID_SIZE << std::endl;
    std::cout << "障碍物占比: " << OBSTACLE_RATIO * 100 << "%" << std::endl;

    // 1. 生成网格
    std::vector<std::vector<int>> grid;
    generateGrid(grid, rng);

    int free_count = 0, obs_count = 0;
    for (const auto& row : grid) {
        for (int cell : row) {
            if (cell == 0) free_count++;
            else obs_count++;
        }
    }
    std::cout << "空闲格子: " << free_count << ", 障碍物: " << obs_count << std::endl;

    // 2. 生成起点和终点
    Point start, goal;
    generateStartGoal(grid, start, goal, rng);
    std::cout << "起点 (Start):  (" << start.row << ", " << start.col << ")" << std::endl;
    std::cout << "终点 (Goal):   (" << goal.row << ", " << goal.col << ")" << std::endl;
    std::cout << "欧几里得距离: "
              << std::hypot(start.row - goal.row, start.col - goal.col) << std::endl;

    // 3. 保存文本网格
    saveGridText(grid, start, goal, "output/grid.txt");

    // 4. 保存 PPM 图片
    saveGridPPM(grid, start, goal, "output/grid_map.ppm");

    std::cout << "\n可使用以下命令查看图片:" << std::endl;
    std::cout << "  open output/grid_map.ppm    (macOS 预览)" << std::endl;
    std::cout << "  python3 -c \"from PIL import Image; Image.open('output/grid_map.ppm').show()\"" << std::endl;

    return 0;
}

File: src/planner/generate_grid.py
import matplotlib.pyplot as plt
import numpy as np
import os
import sys

# 设置字体
plt.rcParams['font.sans-serif'] = ['DejaVu Sans', 'Arial Unicode MS', 'Liberation Sans']
plt.rcParams['axes.unicode_minus'] = True

# ========================== 参数配置 ==========================
GRID_SIZE = 256                # 网格大小 256x256
OBSTACLE_RATIO = 0.2           # 障碍物占比 (0.0 ~ 1.0)
OUTPUT_DIR = 'output'          # 输出目录
OUTPUT_GRID_FILE = 'grid.txt'  # 网格数据文件名

# ========================== 随机种子 (可选固定) ==========================
# 取消下面一行注释可固定随机种子，便于复现
# np.random.seed(42)


def generate_grid(size, obstacle_ratio):
    """生成 size x size 的 Occupancy Grid Map (0=空闲, 1=障碍)。
       使用矩形块 + 墙壁，让地图结构更自然。"""
    grid = np.zeros((size, size), dtype=np.int8)
    target_obs = int(size * size * obstacle_ratio)
    placed = 0

    # 随机放置矩形块
    while placed < target_obs:
        w = np.random.randint(4, 31)
        h = np.random.randint(4, 31)
        r = np.random.randint(0, size)
        c = np.random.randint(0, size)

        r_end = min(r + h - 1, size - 1)
        c_end = min(c + w - 1, size - 1)

        # 跳过重叠超过 30% 的
        area = (r_end - r + 1) * (c_end - c + 1)
        already = int(np.sum(grid[r:r_end+1, c:c_end+1]))
        if already > area * 0.3:
            continue

        grid[r:r_end+1, c:c_end+1] = 1
        placed += area - already

    # 额外添加细长墙壁（横向 + 纵向）
    walls = size // 16
    for i in range(walls):
        length = np.random.randint(20, 81)
        r = np.random.randint(5, size - 5)
        c = np.random.randint(5, size - 5)

        if i % 2 == 0:
            # 横向墙 1~2 格厚
            thick = np.random.randint(1, 3)
            r_end = min(r + thick - 1, size - 1)
            c_end = min(c + length, size - 1)
            grid[r:r_end+1, c:c_end+1] = 1
        else:
            # 纵向墙 1~2 格厚
            thick = np.random.randint(1, 3)
            r_end = min(r + length, size - 1)
            c_end = min(c + thick - 1, size - 1)
            grid[r:r_end+1, c:c_end+1] = 1

    return grid


def generate_start_goal(grid, margin=10):
    """在空闲区域随机选取起点和终点，保证两者不重叠且有一定间距。"""
    size = grid.shape[0]
    free_rc = np.argwhere(grid == 0)

    # 过滤边界区域
    mask = ((free_rc[:, 0] >= margin) & (free_rc[:, 0] < size - margin) &
            (free_rc[:, 1] >= margin) & (free_rc[:, 1] < size - margin))
    valid = free_rc[mask]

    if len(valid) < 2:
        print("错误: 空闲格子不足，无法放置起点和终点!")
        sys.exit(1)

    min_dist = size * 0.3
    while True:
        idx = np.random.choice(len(valid), 2, replace=False)
        start = tuple(valid[idx[0]])
        goal  = tuple(valid[idx[1]])
        if np.linalg.norm(np.array(start) - np.array(goal)) >= min_dist:
            return start, goal


def save_grid(grid, start, goal, filepath):
    """将网格、起点、终点保存为文本文件，供 C++ A* 读取。"""
    size = grid.shape[0]
    with open(filepath, 'w') as f:
        f.write(f"# A* Grid Map\n")
        f.write(f"# size: {size}x{size}\n")
        f.write(f"# start: ({start[0]}, {start[1]})\n")
        f.write(f"# goal:  ({goal[0]}, {goal[1]})\n")
        f.write(f"# 0=free, 1=obstacle\n")
        f.write(f"{size}\n")
        f.write(f"{start[0]} {start[1]}\n")
        f.write(f"{goal[0]} {goal[1]}\n")
        for row in grid:
            f.write(' '.join(str(c) for c in row) + '\n')
    print(f"网格数据已保存至: {filepath}")


def visualize_grid(grid, start, goal, save_path=None):
    """可视化网格，起点蓝色、终点红色。"""
    fig, ax = plt.subplots(figsize=(8, 8))

    cmap = plt.get_cmap('gray_r', 2)
    ax.imshow(grid, cmap=cmap, origin='upper', interpolation='none')

    # 小网格显示格线
    if GRID_SIZE <= 64:
        for x in range(grid.shape[1] + 1):
            ax.axvline(x - 0.5, color='black', linewidth=0.3, alpha=0.3)
        for y in range(grid.shape[0] + 1):
            ax.axhline(y - 0.5, color='black', linewidth=0.3, alpha=0.3)

    # 起点 (蓝色圆)
    ax.plot(start[1], start[0], 'o', color='blue', markersize=10,
            markeredgecolor='darkblue', markeredgewidth=1.5, label='Start')
    # 终点 (红色方)
    ax.plot(goal[1], goal[0], 's', color='red', markersize=10,
            markeredgecolor='darkred', markeredgewidth=1.5, label='Goal')

    ax.set_title(f'A* Grid Map ({grid.shape[0]}x{grid.shape[1]})\n'
                 f'Start: ({start[0]}, {start[1]})  Goal: ({goal[0]}, {goal[1]})',
                 fontsize=14)
    ax.set_xlabel('Column')
    ax.set_ylabel('Row')
    ax.legend(loc='upper right')
    ax.set_xlim(-0.5, grid.shape[1] - 0.5)
    ax.set_ylim(grid.shape[0] - 0.5, -0.5)

    plt.tight_layout()

    if save_path:
        plt.savefig(save_path, dpi=150, bbox_inches='tight')
        print(f"图片已保存至: {save_path}")

    plt.show()


def main():
    print(f"=== A* 网格地图生成 (Python) ===")
    print(f"网格大小: {GRID_SIZE}x{GRID_SIZE}")
    print(f"障碍物占比: {OBSTACLE_RATIO*100:.0f}%")

    # 1. 生成网格
    grid = generate_grid(GRID_SIZE, OBSTACLE_RATIO)
    print(f"空闲格子: {np.sum(grid == 0)}, 障碍物: {np.sum(grid == 1)}")

    # 2. 生成起点和终点
    start, goal = generate_start_goal(grid)
    print(f"起点 (Start):  ({start[0]}, {start[1]})")
    print(f"终点 (Goal):   ({goal[0]}, {goal[1]})")
    print(f"欧几里得距离: {np.linalg.norm(np.array(start) - np.array(goal)):.2f}")

    # 3. 保存数据
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    save_grid(grid, start, goal, os.path.join(OUTPUT_DIR, OUTPUT_GRID_FILE))

    # 4. 可视化
    visualize_grid(grid, start, goal, save_path=os.path.join(OUTPUT_DIR, 'grid_map.png'))


if __name__ == '__main__':
    main()
